aip/0166.md

```diff
@@ -0,0 +1,78 @@
+---
+aip:
+  id: 166
+  state: reviewing
+  created: 2020-06-24
+permalink: /166
+redirect_from:
+  - /0166
+---
+
+# Hierarchically-derived resources
+
+Services occasionally have resources where the resource's effective values are
+determined by consulting both configuration on the resource as well as
+configuration elsewhere, such as a parent in the hierarchy, and "flattening"
+them together. This is most common for configuration resources, which could be
+applied against a resource or a container.
+
+In these situations, users need to distinguish between properties applied to
+the resource directly and properties that are inherited.
+
+## Guidance
+
+Services **should** return only field values on the resource when responding to
+`Get` and `List` methods by default.
+
+Additionally, services **may** add a field to `Get` and `List` requests that
+allows the user to opt-in to receiving the effective configuration instead:
+
+```proto
+message GetBookRequest {
+  // The name of the book.
+  string name = 1 [(google.api.resource_reference) = {
+    type: "library.googleapis.com/Book"
+  }];
+
+  // If true, values that are set on other objects in the hierachy and
+  // inherited by the book will be included.
+  // If false (default), only fields directly set on the book are included.
+  bool effective = 2;
+}
+```
+
+Services **must not** add this field for cases where determining the effective
```

Comment:
I think this guidance should discuss what happens when the consistency guarantees differ between loading a resource and an effective resource which considers inheritance.

Many policy systems actually make it extremely cheap to load an effective policy, however it's eventually consistent as opposed to loading a policy set on a specific resource which would be expected to be strongly consistent (eg., the resource lives on a management plane, but the effective policy/resource is derived from a data/control plane).

Reply:
This makes sense. It does pose some risk, though -- e.g. declarative tooling does not play nicely with eventual consistency. Is it guaranteed to be the case that the values _on the resource being queried_ are strongly consistent (and just the inherited ones are eventually consistent)? That is my intuition on what the guidance should be.